### ISSUE TITLE
Bad result

### ISSUE THREAD

**Issue by ghost** [mondokm/ltlverif#1]
**Note:** the code has changed, that might be one source of the issue! The question is more of a "is the LTL in bad format?"

LTL: neither "FG (not err)" nor "XG (not err)" work as expected: both return "LTL Expression holds" for both CFAs below:

CFA files:
(changed from counter5_true.cfa)
```CFA
main process cfa {
    var x : int
    var err : bool

    init loc IL1
    loc L0
    loc L1
    loc L2
    loc L3
    final loc END
    loc ERR
    error loc ERR1

    IL1 -> L0 { err := false }
    L0 -> L1 { x := 0 }
    L1 -> L2 { assume x < 5 }
    L1 -> L3 { assume not (x < 5) }
    L2 -> L1 { x := x + 1 }
    L3 -> END { assume x <= 5 }
    L3 -> ERR { assume not (x <= 5) }
    ERR -> ERR1 { err := true }
}
```

(changed from counter5_false.cfa)
```CFA
main process cfa {
    var x : int
    var err : bool

    loc L0
    init loc IL1
    loc L1
    loc L2
    loc L3
    final loc END
    loc ERR
    error loc ERR1

    IL1 -> L0 { err := false }
    L0 -> L1 { x := 0 }
    L1 -> L2 { assume x < 5 }
    L1 -> L3 { assume not (x < 5) }
    L2 -> L1 { x := x + 1 }
    L3 -> END { assume x <= 5 }
    L3 -> ERR { assume not (x <= 5) }
    ERR -> ERR1 { err := true }
}
```

**Comment by ghost** [comment]
The changes *should not* affect the result... But I am not sure enough.

The generated Buchi automaton:
```Buchi
HOA: v1
name: "!ap0"
States: 3
Start: 1
AP: 1 "ap0"
acc-name: Buchi
Acceptance: 1 Inf(0)
properties: trans-labels explicit-labels state-acc complete
properties: deterministic stutter-invariant terminal
--BODY--
State: 0 {0}
[t] 0
State: 1
[!0] 0
[0] 2
State: 2
[t] 2
--END--
```

From Spot's online converter (I guess the same tool, maybe this helps?):
![image](https://user-images.githubusercontent.com/68019200/87794815-7cc8f780-c847-11ea-83a8-5da65cdc410f.png)

**Comment by ghost** [comment]
```Java
    @Override
    public boolean isFinalState(ProductState prod) {
        return ((CfaProductState)prod).getLoc().equals(cfa.getFinalLoc());
    }
```

to

```Java
    @Override
    public boolean isFinalState(ProductState prod) {
        return ((CfaProductState)prod).getLoc().equals(cfa.getFinalLoc()) ||
               ((CfaProductState)prod).getLoc().equals(cfa.getErrorLoc());
    }
```

in CfaSUT.java did not change output. (LTL expression holds on both CFAs for both `XG(not err)` and `FG(not err)`)

**Comment by ghost** [comment]
Changing `XG (not err)` to `X G (not err)` fixed half of the problems, will update issue soon...

**Comment by mondokm** [comment]
You're correct, a space is required between two temporal operators.
Support for temporal operator `X` was removed, because it can't be evaluated in the final state of finite paths. When we discussed this issue last year, two possible solutions came up. The first option would limit the scope of expressions to LTL\X, which is LTL without X. The second option would treat final states which normally have no loops as if they had a loop edge and no outgoing edges. This second option was introduced for 'final' states, but support for the 'X' operator wasn't restored.
